substrate/frame/migrations/src/mock.rs

```diff
@@ -0,0 +1,297 @@
+///
+/// It achieves this by using a thread-local storage to store the migrations to run.
+pub struct MigrationsStorage;
+impl SteppedMigrations for MigrationsStorage {
+	fn len() -> u32 {
+		MIGRATIONS.with(|m| m.borrow().len()) as u32
+	}
+
+	fn nth_id(n: u32) -> Option<Vec<u8>> {
+		let k = MIGRATIONS.with(|m| m.borrow().get(n as usize).map(|k| *k));
+		k.map(|(kind, steps)| mocked_id(kind, steps).into_inner())
+	}
+
+	fn nth_step(
+		n: u32,
+		cursor: Option<Vec<u8>>,
+		_meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>> {
+		Some(MIGRATIONS.with(|m| {
+			let (kind, steps) = m.borrow()[n as usize];
+
+			let mut count: u32 =
+				cursor.as_ref().and_then(|c| Decode::decode(&mut &c[..]).ok()).unwrap_or(0);
+			log::debug!("MockedMigration: Step {}", count);
+			if count != steps || matches!(kind, TimeoutAfter) {
+				count += 1;
+				return Ok(Some(count.encode().try_into().unwrap()))
+			}
+
+			match kind {
+				SucceedAfter => {
+					log::debug!("MockedMigration: Succeeded after {} steps", count);
+					Ok(None)
+				},
+				HightWeightAfter(required) => {
+					log::debug!("MockedMigration: Not enough weight after {} steps", count);
+					Err(SteppedMigrationError::InsufficientWeight { required })
+				},
+				FailAfter => {
+					log::debug!("MockedMigration: Failed after {} steps", count);
+					Err(SteppedMigrationError::Failed)
+				},
+				TimeoutAfter => unreachable!(),
+			}
+		}))
+	}
+
+	fn nth_transactional_step(
+		n: u32,
+		cursor: Option<Vec<u8>>,
+		meter: &mut WeightMeter,
+	) -> Option<Result<Option<Vec<u8>>, SteppedMigrationError>> {
+		// FAIL-CI
+		Self::nth_step(n, cursor, meter)
+	}
+
+	fn nth_max_steps(n: u32) -> Option<Option<u32>> {
+		MIGRATIONS
+			.with(|m| m.borrow().get(n as usize).map(|s| *s))
+			.map(|(_, s)| Some(s))
+	}
+
+	fn cursor_max_encoded_len() -> usize {
+		65_536
+	}
+
+	fn identifier_max_encoded_len() -> usize {
+		256
+	}
+}
+
+impl MigrationsStorage {
+	/// Set the migrations to run.
+	pub fn set(migrations: Vec<(MockedMigrationKind, u32)>) {
+		MIGRATIONS.with(|m| *m.borrow_mut() = migrations);
+	}
+}
+
+impl crate::Config for Test {
+	type RuntimeEvent = RuntimeEvent;
+	type Migrations = MigrationsStorage;
+	type CursorMaxLen = ConstU32<65_536>;
+	type IdentifierMaxLen = ConstU32<256>;
+	type OnMigrationUpdate = MockedOnMigrationUpdate;
+	type ServiceWeight = ServiceWeight;
+	type WeightInfo = ();
+}
+
+// Build genesis storage according to the mock runtime.
+pub fn new_test_ext() -> sp_io::TestExternalities {
+	sp_io::TestExternalities::new(Default::default())
+}
+
+/// Run this closure in test externalities.
+pub fn test_closure<R>(f: impl FnOnce() -> R) -> R {
+	let mut ext = new_test_ext();
+	ext.execute_with(f)
+}
+
+pub fn run_to_block(n: u32) {
+	while System::block_number() < n as u64 {
+		if System::block_number() > 1 {
+			Migrations::on_finalize(System::block_number());
+			System::on_finalize(System::block_number());
```

Comment:
I thought `on_finalize` is called after migrations every block?

Reply:
Yea I moved it down now. Normally we do it the other way around (i think for the genesis case) but here it does not matter.